### PR TITLE
feat: allow variable tensor dimension

### PR DESCRIPTION
**Pull request by Jackmin801**
Goals:

- Implements https://github.com/docarray/docarray/issues/871
- [x] Implement for TorchTensor
- [x] Implement for NdArray
- [x] check and update documentation, if required. See [guide](https://github.com/jina-ai/jina/blob/master/CONTRIBUTING.md#documentation-guidelines)

### CODE REVIEW

**Comment by Jackmin801** [comment]
Do we want to add an example usage in the docstrings?

**Review by JohannesMessner — CHANGES_REQUESTED** [review]
Can this logic be move into the `AbstractTensor` class so that it "just works" when adding a new tensor type? Also I'm guessing variables (`'x'`) are not shared between tensors (or the same Document), right?

**Comment by JohannesMessner** [comment]
> Do we want to add an example usage in the docstrings?

Yes please!

File: docarray/typing/tensor/ndarray.py [hunk 1]
```diff
@@ -83,9 +87,29 @@ def __get_validators__(cls):
     def __docarray_validate_shape__(cls, t: T, shape: Tuple[int]) -> T:  # type: ignore
         if t.shape == shape:
             return t
+        elif any(isinstance(dim, str) for dim in shape):
+            if len(t.shape) != len(shape):
+                raise ValueError(
+                    f"Array shape mismatch. Expected {shape}, got {t.shape}"
```
Comment (samsja) [inline docarray/typing/tensor/ndarray.py:93]: single quote

File: docarray/typing/tensor/ndarray.py [hunk 2]
```diff
@@ -83,9 +87,29 @@ def __get_validators__(cls):
     def __docarray_validate_shape__(cls, t: T, shape: Tuple[int]) -> T:  # type: ignore
```
Comment (samsja) [inline docarray/typing/tensor/ndarray.py:87]: shape can be Tuple[Union[int, str]]

File: docarray/typing/tensor/ndarray.py [hunk 3]
```diff
@@ -83,9 +87,29 @@ def __get_validators__(cls):
...
+        elif any(isinstance(dim, str) for dim in shape):
+            if len(t.shape) != len(shape):
+                raise ValueError(
+                    f"Array shape mismatch. Expected {shape}, got {t.shape}"
+                )
+            known_dims: Dict[str, int] = {}
+            for tdim, dim in zip(t.shape, shape):
+                if isinstance(dim, int) and tdim != dim:
+                    raise ValueError(
+                        f"Array shape mismatch. Expected {shape}, got {t.shape}"
```
Comment (samsja) [inline docarray/typing/tensor/ndarray.py:99]: single quote

File: docarray/typing/tensor/ndarray.py [hunk 4]
```diff
@@ -83,9 +87,29 @@ def __get_validators__(cls):
...
+                    )
+                elif isinstance(dim, str):
+                    if dim in known_dims and known_dims[dim] != tdim:
+                        raise ValueError(
+                            f"Array shape mismatch. Expected {shape}, got {t.shape}"
```
Comment (samsja) [inline docarray/typing/tensor/ndarray.py:104]: single quote

**Review by samsja — COMMENTED** [review]
would be nice to find a way to factorize the code between torch and numpy

**Comment by samsja** [comment]
> First question, unfortunately not. Numpy and torch have slightly different api, particularly in this case, in the view vs reshape. And yep, the 'x' isn't shared.

cannot we abstract the reshaping part, somewhere ?

**Comment by JohannesMessner** [comment]
> > First question, unfortunately not. Numpy and torch have slightly different api, particularly in this case, in the view vs reshape. And yep, the 'x' isn't shared.
> 
> cannot we abstract the reshaping part, somewhere in the computational backend ?

I suggest the same

File: docarray/typing/tensor/abstract_tensor.py [hunk 5]
```diff
@@ -60,26 +72,43 @@ class AbstractTensor(Generic[ShapeT], AbstractType, ABC):
...
-        - If the shape of `t` is not equal to `shape`,
-            but can be reshaped to `shape`, return `t` reshaped to `shape`.
-        - If the shape of `t` is not equal to `shape`
-            and cannot be reshaped to `shape`, raise a ValueError.
-
-        :param t: The tensor to validate.
-        :param shape: The shape to validate against.
-        :return: The validated tensor.
-        """
-        ...
```
Comment (samsja) [inline docarray/typing/tensor/abstract_tensor.py:82]: lets keep the docstring

File: docarray/typing/tensor/abstract_tensor.py [hunk 6]
```diff
@@ -60,26 +72,43 @@ class AbstractTensor(Generic[ShapeT], AbstractType, ABC):
...
+    def __docarray_validate_shape__(
+        cls, t: T, shape: Tuple[Union[int, str]]
+    ) -> T:  # type: ignore
```
Comment (samsja) [inline docarray/typing/tensor/abstract_tensor.py:77]: why is it type ignore her e?

File: docarray/typing/tensor/abstract_tensor.py [hunk 7]
```diff
@@ -60,26 +72,43 @@ class AbstractTensor(Generic[ShapeT], AbstractType, ABC):
...
+        if t.shape == shape:
```
Comment (samsja) [inline docarray/typing/tensor/abstract_tensor.py:78]: Here can we assume that t has a shape field ? I am not sure maybe it is better to call the `get_comp_backend`

File: docarray/typing/tensor/abstract_tensor.py [hunk 8]
```diff
@@ -60,26 +72,43 @@ class AbstractTensor(Generic[ShapeT], AbstractType, ABC):
...
+                        known_dims[dim] = tdim
+            else:
+                return t
+        else:
+            warnings.warn(
+                f'Tensor shape mismatch. Reshaping tensor '
+                f'of shape {t.shape} to shape {shape}'
+            )
+            try:
+                value = cls.__docarray_from_native__(t.reshape(shape))
```
Comment (samsja) [inline docarray/typing/tensor/abstract_tensor.py:106]: same here I am not sure we can do reshape directly

File: docarray/typing/tensor/abstract_tensor.py (hunk 6 above)
Comment (Jackmin801) [reply docarray/typing/tensor/abstract_tensor.py:77]: Ah its an artifact from copying the signature in TorchTensor. Seems to work fine without it. Removed

File: docarray/typing/tensor/abstract_tensor.py [hunk 9]
```diff
@@ -213,3 +214,18 @@ def _docarray_to_json_compatible(self):
...
+        :return: a tensor with the same data and number of elements as self
+            but with the specified shape.
+        """
+        ...
+
+    @property
+    @abc.abstractmethod
+    def shape(self) -> Tuple[int, ...]:
+        """The shape of this tensor."""
+        ...
```
Comment (samsja) [inline docarray/typing/tensor/abstract_tensor.py:231]: we should move these to the computation backend IMO

File: docarray/typing/tensor/abstract_tensor.py (hunk 7 above)
Comment (Jackmin801) [reply docarray/typing/tensor/abstract_tensor.py:78]: Hrmm ok. Ill do that

File: docarray/typing/tensor/abstract_tensor.py (hunk 5 above)
Comment (Jackmin801) [reply docarray/typing/tensor/abstract_tensor.py]: okie

File: docarray/typing/tensor/abstract_tensor.py (hunk 9 above)
Comment (Jackmin801) [reply docarray/typing/tensor/abstract_tensor.py]: okei

File: docarray/computation/torch_backend.py [hunk 10]
```diff
@@ -73,6 +73,50 @@ def none_value() -> Any:
...
+        Gives a new shape to tensor without changing its data.
+
+        :param tensor: tensor to be reshaped
+        :param shape: the new shape
+        :return: a tensor with the same data and number of elements as tensor
+            but with the specified shape.
+        """
+        ...
+
+    @overload
```
Comment (samsja) [inline docarray/computation/torch_backend.py:93]: Would be nice to automatically generate the overload. But lets not do it in this PR unless you found a brillant solution haha

Comment (Jackmin801) [reply docarray/computation/torch_backend.py:93]: I dont think it works :( https://github.com/python/mypy/issues/11488

Comment (samsja) [reply docarray/computation/torch_backend.py:93]: but we could generate the code though

Comment (samsja) [reply docarray/computation/torch_backend.py:93]: but yeah lets see later

Comment (samsja) [reply docarray/computation/torch_backend.py:93]: nah not code generation via our friend the bot.

But we could have a helper function that generate 3 methods with the correct overload

Comment (JohannesMessner) [reply docarray/computation/torch_backend.py:93]: could it be done via decorator somehow?

**Review by JohannesMessner — APPROVED** [review]

**Review by samsja — APPROVED** [review]

**Pull request merged**